File: creek/src/main/scala/org/creek/collection/Graph.scala
package org.creek.collection

import scala.collection.mutable
import scala.collection.mutable.ListBuffer

// TODO: make the API with Option for getters, and Try for setters
// TODO: update to better API // add/delete edges, updateNode ... and so on
// TODO: add notes on adding an empty edge
// TODO: scalaDocs: N must implement HashCode and Equals
// TODO: reverse graph // meaningful for directed graph
// TODO: Kosaraju-Sharir for Strongly connected components - Sedgewick
// TODO: add ordering for keys, and use TreeMap??
// Or this one ??? Gabow's algorithm (aka Cheriyan-Mehlhorn algorithm)
trait Graph[N]{
  def getNodes : List[N]
  def addEdge(from: N, to: N, weight: Double = 0.0)
  def adjacentNodes(node: N) : List[N]
  def adjacentEdges(node: N) : List[Edge[N]]
  def size : Int
}

object Graph {
  def apply[N](kind: GraphType) : Graph[N] = {
    kind match {
      case Undirected => new UndirectedGraph[N]
      case Directed => new DirectedGraph[N]
    }
  }
}

sealed trait GraphType
case object Undirected extends GraphType
case object Directed extends GraphType

class Edge[N] private[collection](val from: N, val to: N, val weight: Double = 0.0){
  override def toString: String = s"$from - $to : $weight"
}

private abstract class AbstractGraph[N] extends Graph[N]{

  protected val nodes: mutable.Map[N, ListBuffer[Edge[N]]] = mutable.Map[N, ListBuffer[Edge[N]]]()

  protected def addNode(node: N): Unit = if(nodes.contains(node) == false) nodes += (node -> ListBuffer[Edge[N]]())
  override def getNodes: List[N] = nodes.keySet.toList
  override def adjacentNodes(node: N): List[N] = adjacentEdges(node).map(edge => edge.to)
  override def adjacentEdges(node: N): List[Edge[N]] = {
    nodes.get(node) match {
      case Some(value) => value.toList
      case None => List[Edge[N]]()
    }
  }
  override def size: Int = nodes.size
}

private class UndirectedGraph[N] extends AbstractGraph[N] {
  override def addEdge(from: N, to: N, weight: Double = 0.0): Unit = {
    if(from == null) return
    addNode(from)
    if(to == null) return // no edge, just a single node
    addNode(to)
    nodes.getOrElseUpdate(from, ListBuffer[Edge[N]]()) += new Edge[N](from, to, weight)
    nodes.getOrElseUpdate(to, ListBuffer[Edge[N]]()) += new Edge[N](to, from, weight)
  }
}

private class DirectedGraph[N] extends AbstractGraph[N] {
  override def addEdge(from: N, to: N, weight: Double = 0.0): Unit = {
    if(from == null) return
    addNode(from)
    if(to == null) return // no edge, just a single node
    addNode(to)
    nodes.getOrElseUpdate(from, ListBuffer[Edge[N]]()) += new Edge[N](from, to, weight)
  }
}

File: creek_java/src/io/creek/Graph.java
package io.creek;

import static io.creek.Graph.GraphType.DIRECTED;
import static io.creek.Graph.GraphType.UNDIRECTED;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Graph interface.</br>
 * {@code Graph<String> graph = Graph.createGraph(GraphType.UNDIRECTED);}
 * 
 * @param <T> the graph type
 */
public interface Graph<T> {

	List<T> getNodes();

	List<T> getAdjacentNodes(T node);

	List<Edge<T>> getAdjacentEdges(T node);

	void addEdge(T from, T to, double weight);

	void addEdge(T from, T to);

	GraphType getType();

	/** create {@link Graph} instance */
	static <T> Graph<T> createGraph(GraphType type) {
		switch (type) {
		case DIRECTED:
			return new DirectedGraph<T>();
		case UNDIRECTED:
			return new UndirectedGrpah<T>();
		default:
			return null;
		}
	}

	public enum GraphType {
		DIRECTED, UNDIRECTED;
	}

	abstract class AbstractGraph<T> implements Graph<T> {

		private Map<T, List<Edge<T>>> nodes = new HashMap<>();

		protected void addNode(T node) {
			nodes.putIfAbsent(node, new ArrayList<>());
		}

		protected void addEdge(T node, Edge<T> edge) {
			nodes.computeIfAbsent(node, list -> new ArrayList<>()).add(edge);
		}

		@Override
		public List<T> getNodes() {
			return new ArrayList<>(nodes.keySet().stream().collect(Collectors.toList()));
		}

		@Override
		public List<Edge<T>> getAdjacentEdges(T node) {
			return nodes.getOrDefault(node, new ArrayList<>());
		}

		@Override
		public List<T> getAdjacentNodes(T node) {
			return getAdjacentEdges(node).stream().filter(edge -> edge.to != null).map(edge -> edge.to)
					.collect(Collectors.toList());
		}

		@Override
		public void addEdge(T from, T to) {
			addEdge(from, to, 0.0);
		}
	}

	class DirectedGraph<T> extends AbstractGraph<T> {

		@Override
		public void addEdge(T from, T to, double weight) {
			if (from != null) addNode(from);
			if (to != null) addNode(to);
			if (from != null && to != null) addEdge(from, new Edge<T>(from, to, weight));
		}

		@Override
		public GraphType getType() {
			return DIRECTED;
		}
	}

	class UndirectedGrpah<T> extends AbstractGraph<T> {

		@Override
		public void addEdge(T from, T to, double weight) {
			if (from != null) addNode(from);
			if (to != null) addNode(to);
			if (from != null && to != null) {
				addEdge(from, new Edge<T>(from, to, weight));
				addEdge(to, new Edge<T>(to, from, weight));
			}
		}

		@Override
		public GraphType getType() {
			return UNDIRECTED;
		}
	}

	public static class Edge<T> {
		private T from;
		private T to;
		private double weight;

		private Edge(T from, T to, double weight) {
			super();
			this.from = from;
			this.to = to;
			this.weight = weight;
		}

		public T from() {
			return from;
		}

		public T to() {
			return to;
		}

		public double weight() {
			return weight;
		}

		@Override
		public String toString() {
			return "[ " + from + " - " + to + " : " + weight + "]";
		}
	}
}
